cosmwasm/contracts/wormchain-accounting/src/contract.rs

```diff
@@ -0,0 +1,491 @@
+use std::marker::PhantomData;
+
+use accounting::{
+    query_balance, query_modification, query_transfer,
+    state::{account, transfer, Modification, TokenAddress, Transfer},
+};
+use anyhow::{ensure, Context};
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    from_binary, to_binary, Binary, ConversionOverflowError, CosmosMsg, Deps, DepsMut, Empty, Env,
+    Event, MessageInfo, Order, Response, StdError, StdResult, Uint256, WasmMsg,
+};
+use cw2::set_contract_version;
+use cw_storage_plus::Bound;
+use tinyvec::{Array, TinyVec};
+use wormhole::token::Message;
+use wormhole_bindings::WormholeQuery;
+
+use crate::{
+    bail,
+    error::{AnyError, ContractError},
+    msg::{
+        AllAccountsResponse, AllModificationsResponse, AllPendingTransfersResponse,
+        AllTransfersResponse, ExecuteMsg, Instantiate, InstantiateMsg, MigrateMsg, Observation,
+        QueryMsg, Upgrade,
+    },
+    state::{self, Data, PendingTransfer, PENDING_TRANSFERS, TOKENBRIDGE_ADDR},
+};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:wormchain-accounting";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut<WormholeQuery>,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, AnyError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)
+        .context("failed to set contract version")?;
+
+    let signatures = msg.signatures.iter().cloned().map(From::from).collect();
+    let _: Empty = deps
+        .querier
+        .query(
+            &WormholeQuery::VerifyQuorum {
+                data: msg.instantiate.clone(),
+                observer_set_index: msg.observer_set_index,
+                signatures,
+                block_time: env.block.height,
+            }
+            .into(),
+        )
+        .context(ContractError::VerifyQuorum)?;
+
+    let init: Instantiate =
+        from_binary(&msg.instantiate).context("failed to parse `Instantiate` message")?;
+
+    let tokenbridge_addr = deps
+        .api
+        .addr_validate(&init.tokenbridge_addr)
+        .context("failed to validate tokenbridge address")?;
+
+    TOKENBRIDGE_ADDR
+        .save(deps.storage, &tokenbridge_addr)
+        .context("failed to save tokenbridge address")?;
+
+    let event =
+        accounting::instantiate(deps, init.into()).context("failed to instantiate accounting")?;
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute("owner", info.sender)
+        .add_event(event))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(_deps: DepsMut<WormholeQuery>, _env: Env, _msg: MigrateMsg) -> StdResult<Response> {
+    Ok(Response::default())
```

Comment:
I believe as is the cosmos account owner would be freely available to change the contract.

My proposal would be to require a VAA assertion before allowing the contract to upgrade.

Reply:
I believe `migrate` can be freely executed by the owner of the contract which we don't want.  Maybe move the VAA verification currently in `upgrade_contract` to `migrate`?

Reply:
I believe what we do for other cosmwasm contracts is to set the owner to itself.  Then upgrades can only happen through governance and no one else should be able to call `migrate`.

If we can't set the contract owner to itself then the `UpgradeContract` message would be useless since the contract would not be allowed to call migrate.  This actually caused some issues during testing where I had to do a gross hack to be able to set the contract owner to itself (see proper_instantiate() in tests/helpers/mod.rs).

Reply:
Ah okay that makes sense, thanks for explaining.